### PR TITLE
Fix (some) n+1/eager loading issues

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

I ran almost all of the tests (except system tests) using `BULLET=true` and fixed (most of) the issues I've encountered

Some are n+1, some are unnecessary eager loadings

### CODE REVIEW

File: app/controllers/additional_content_boxes_controller.rb [hunk 1]
```diff
@@ -6,16 +6,18 @@ class AdditionalContentBoxesController < ApplicationController
...
+
     if (!user_signed_in? || params[:state] == "include_sponsors") &&
         @article.user.permit_adjacent_sponsors &&
         randomize
-      @boosted_article = Suggester::Articles::Boosted.new(
-        (@article.decorate.cached_tag_list_array + @article.boosted_additional_tags.split).sample,
-        not_ids: (article_ids + @suggested_articles.pluck(:id)), area: "additional_articles",
-      ).suggest
+
+      tag = (@article.decorate.cached_tag_list_array + @article.boosted_additional_tags.split).sample
```
Comment (rhymes) [inline app/controllers/additional_content_boxes_controller.rb]: just a cleanup

File: app/controllers/dashboards_controller.rb [hunk 2]
```diff
@@ -16,12 +16,17 @@ def show
     if params[:which] == "organization" && params[:org_id] && (@user.org_admin?(params[:org_id]) || @user.any_admin?)
       target = @organizations.find_by(id: params[:org_id])
       @organization = target
+      @articles = target.articles
+    else
+      # if the target is a user, we need to eager load the organization
+      @articles = target.articles.includes(:organization)
```
Comment (rhymes) [inline app/controllers/dashboards_controller.rb]: here we were eager loading the organization even when the target was an organization, so I removed that

File: app/controllers/dashboards_controller.rb [hunk 3]
```diff
@@ -16,12 +16,17 @@ def show
...
     end
 
-    @articles = target.articles.includes(:organization).sorting(params[:sort]).decorate
+    @articles = @articles.sorting(params[:sort]).decorate
 
     # Updates analytics in background if appropriate
-    Articles::UpdateAnalyticsJob.perform_later(current_user.id) if @articles && ApplicationConfig["GA_FETCH_RATE"] < 50 # Rate limit concerned, sometimes we throttle down.
+    update_analytics = @articles && ApplicationConfig["GA_FETCH_RATE"] < 50 # Rate limited, sometimes we throttle down
+    Articles::UpdateAnalyticsJob.perform_later(current_user.id) if update_analytics
```
Comment (rhymes) [inline app/controllers/dashboards_controller.rb]: easier to read

File: app/controllers/notifications_controller.rb [hunk 4]
```diff
@@ -26,12 +26,20 @@ def index
...
     @notifications = @notifications.where("notified_at < ?", notified_at_offset) if notified_at_offset
 
-    @notifications = NotificationDecorator.decorate_collection(@notifications.limit(num))
+    @notifications = @notifications.limit(num)
+
+    # after notifications are paginated/limited, we check if any of the notifiables is "Comment" with no action
+    # because its respective view is the only one that actually uses ".notifiable" to render,
+    # this way we save useless eager loadings
+    notifiable_eager_loading_params = { notifiable_type: %w[Comment], action: [nil, ""] }
+    @notifications = @notifications.includes(:notifiable) if @notifications.exists?(notifiable_eager_loading_params)
```
Comment (rhymes) [inline app/controllers/notifications_controller.rb]: This is probably the most controversial change, because it does a `SELECT exists` for each pagination. The reason is that we were performing an eager load (which is a `SELECT *`) even when none of the rendered notifications rendered the "notifiable" object. Here we add a quick `SELECT exists` (which runs on 8 or 30 items), to potentially avoid including data we don't need to render.

File: app/services/article_api_index_service.rb [hunk 5]
```diff
@@ -74,18 +74,29 @@ def tag_articles
   end
 
   def top_articles
-    Article.published.order("positive_reactions_count DESC").where("published_at > ?", top.to_i.days.ago).
+    Article.published.includes(:user, :organization).
```
Comment (rhymes) [inline app/services/article_api_index_service.rb]: we weren't loading those two relations that are accessed in the JSON rendering for the API

File: app/services/article_api_index_service.rb [hunk 6]
```diff
@@ -74,18 +74,29 @@ def tag_articles
...
 
   def state_articles(state)
-    if state == "fresh"
-      Article.published.
-        where("positive_reactions_count < ? AND featured_number > ? AND score > ?", 2, 7.hours.ago.to_i, -2)
-    elsif state == "rising"
-      Article.published.
-        where("positive_reactions_count > ? AND positive_reactions_count < ? AND featured_number > ?", 19, 33, 3.days.ago.to_i)
-    end
+    articles = Article.published.includes(:user, :organization)
```
Comment (rhymes) [inline app/services/article_api_index_service.rb]: we weren't loading those two relations that are accessed in the JSON rendering for the API

File: app/services/article_api_index_service.rb [hunk 7]
```diff
@@ -74,18 +74,29 @@ def tag_articles
...
+               elsif state == "rising"
+                 articles.where(
+                   "positive_reactions_count > ? AND positive_reactions_count < ? AND featured_number > ?",
+                   19, 33, 3.days.ago.to_i
+                 )
+               else
+                 Article.none
+               end
+
+    articles.page(page).per(per_page || DEFAULT_PER_PAGE)
```
Comment (rhymes) [inline app/services/article_api_index_service.rb]: added pagination for fresh or rising articles

File: app/services/article_api_index_service.rb [hunk 8]
```diff
@@ -74,18 +74,29 @@ def tag_articles
...
+                 articles.where(
+                   "positive_reactions_count < ? AND featured_number > ? AND score > ?", 2, 7.hours.ago.to_i, -2
+                 )
+               elsif state == "rising"
+                 articles.where(
+                   "positive_reactions_count > ? AND positive_reactions_count < ? AND featured_number > ?",
+                   19, 33, 3.days.ago.to_i
+                 )
+               else
+                 Article.none
```
Comment (rhymes) [inline app/services/article_api_index_service.rb]: explicitly handles the case of invalid state

File: spec/requests/additional_content_boxes_spec.rb [hunk 9]
```diff
@@ -1,35 +1,35 @@
...
+  let_it_be(:regular_article) { create(:article, user: user, tags: [tag.name]) }
 
   describe "GET /additional_content_boxes" do
     it "returns an article if there is a published/featured one" do
-      suggestion = build_stubbed(:article, published: true, featured: true, path: "blah")
-      suggestion.title = "Title of the article"
-
-      fake = instance_double(Suggester::Articles::Classic)
-      allow(Suggester::Articles::Classic).to receive(:new).and_return(fake)
-      allow(fake).to receive(:get).and_return([suggestion])
```
Comment (rhymes) [inline spec/requests/additional_content_boxes_spec.rb]: because of this mock we weren't actually testing the `get` method, the advantage of building an article that will be found by the "suggester" is that we're sure that the test is actually doing something useful

File: spec/requests/api/v0/articles_spec.rb [hunk 10]
```diff
@@ -167,6 +167,37 @@ def json_response
         expect(json_response.length).to eq(1)
       end
     end
+
+    context "with state param" do
```
Comment (rhymes) [inline spec/requests/api/v0/articles_spec.rb]: added tests for the `state` param which was untested until now

File: app/controllers/notifications_controller.rb (hunk 4 above)
Comment (benhalpern) [reply app/controllers/notifications_controller.rb]: In giving this another look, the only situations we call `.notifiable` at all are in `_article.html.erb` and `_comment.html.erb`.

In `_article...` the code is...

```html
    <div>
      <% if json_data["article"]["published_at"] %>
        <small><%= time_ago_in_words json_data["article"]["published_at"] %> ago</small>
      <% else %>
        <small><%= time_ago_in_words notification.notifiable.published_at %> ago</small>
      <% end %>
    </div>
```

Which must be like this because we added "published at" to notifications later and some might not have this? This conditional can probably safely be removed in favor of the `json_data["article"]["published_at"]` by now as the old ones have faded away.

And then in `_comment...` we have...

```html
      <div>
        <small><%= time_ago_in_words notification.notifiable.created_at %> ago</small>
      </div>
```

Which seems similar to above in that the `json_data` must not include `created_at` so we have to call this, but I think the better solution would be to add this info to the `json_data` itself, and then add a conditional for its existence?

I think that conditional doesn't need to actually render the time if it doesn't exist. I think it could basically be like...

```html
  <% if json_data["comment"]["created_at"] %>
      <div>
        <small><%= time_ago_in_words json_data["comment"]["created_at"] %> ago</small>
      </div>
  <% end %>
```

And `created_at` needs to get added to the `comment_data`. As long as it doesn't mess up the formatting I don't think we need to render anything if `json_data["comment"]["created_at"]` doesn't exist unless it messes with formatting because this isn't super critical info and comment notifications are transient and will soon fade away. We could backfill info if that was needed, but I don't think it's necessary.

TL;DR — We do this, I don't think we need to call the `includes` at all. Seems like the includes only deals with a couple unnecessary exceptions to the general serialization approach.

This is explanation was pretty complicated, does it all make sense @rhymes?

Comment (rhymes) [reply app/controllers/notifications_controller.rb]: It makes perfect sense! I'll modify the article partial and add the date `comment_data`. Do you think it should be in this PR or in a following one?

File: app/controllers/additional_content_boxes_controller.rb (hunk 1 above)
Comment (mstruve) [reply app/controllers/additional_content_boxes_controller.rb]: This is so much more readable! ❤️

File: app/controllers/notifications_controller.rb (hunk 4 above)
Comment (mstruve) [reply app/controllers/notifications_controller.rb]: Why not add it to this PR? I'm also good with a followup since I know you will do it right away 😉

Comment (rhymes) [reply app/controllers/notifications_controller.rb]: I realized it's a bigger change than I anticipated. I have to change `app/services/notifications.rb`, probably the two workers related to sending these two notifications and likely a bunch of tests. I'd feel more comfortable doing this in two steps :D
